Eliminate Input Order Dependency in AirflowNetwork Linkage Objects

RKStrand opened the pull request:
Pull request overview
---------------------

- Fixes #10652

### Description of the purpose of this PR

A user noted that an error message was being produced by EnergyPlus that indicated that some of the input for AirFlowNetwork inputs were not in the correct order.  EnergyPlus has always functioned within the concept of the input can be in any order.  So, this violated that principle.  While there is an internal requirement within AFN that the first encountered Linkage object has a node that is connected to actual E+ HVAC equipment, this can be accommodated without enforcing that the input be in a particular order.  This solution fixes the problem by searching throughout the data that has been read from input (now using JSON directly) until it finds an AirLoopNum that pertains to the air flow network.  In testing, very tiny differences were noted in some sizing results.  A unit test verifies the fix and several existing unit tests that used hard-wired indices for making comparisons were adjusted since the input is read in a slightly different order now.

### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction of Order Dependency Issue in AFN

The input for AirflowNetwork:Distribution:Linkage objects required that the first instance had a node that was an EnergyPlus node rather than just a node within the AFN.  Thus, the input was order dependent--something that E+ says it's not.  This fix reorders the reading of the input object so that the first one it reads is now one that has an EnergyPlus node.  The user input file that previously received a severe/fatal now runs successfully and has output that matches a version of the input file where the objects didn't violate the order issue.

RKStrand committed:
Unit Test and Adjustments to Existing Tests

A unit test was created to test the new integer function that determines the shift in input.  In addition, since the order of the AFN linkage statements is now changed internally, several of the unit tests for AFN had to have hard-wired indices adjusted to reflect the new order.

RKStrand committed:
Corrections of AFN Reordering

Reordering caused some crashes in ci and this revealed that the new subroutine was missing a condition.  This was fixed, the new unit test was modified/strengthened, and the existing unit tests that were changed in the last commit were backed out.

rraustad commented on src/EnergyPlus/AirflowNetwork/src/Solver.cpp:
```diff
@@ -12289,6 +12294,74 @@ namespace AirflowNetwork {
+            }
+        }
+        return shiftResult;
+    }
```
This just looks like it's finding the first AirflowNetwork:Distribution:Linkage that matches one of the AirflowNetwork:Distribution:Node objects and then reading the AirflowNetwork:Distribution:Linkage objects in order. I don't know how that aligns these objects in some manner or how this works for other random orders of AirflowNetwork:Distribution:Linkage objects so I can't comment more. I would ask that @jasondegraw look at this.

jasondegraw reviewed (changes requested):
The issue does appear to be legitimate, and this PR does appear to fix the issue, but it's adding additional array-based input processing. I hate to be "that guy", but this really needs to be done with the JSON-based input processing, otherwise it'll have to be redone later.

RKStrand commented:
@jasondegraw I have a question regarding your comment about this needing to be done in JSON as well.  Now, I'll be the first to admit that I'm not 100% up to speed on JSON yet, so if this question is dumb, at least you'll know why.  Since this fix does not alter in any way reading the data from JSON or the IDF but simply looks at what was contained in the input file, how will the input file format matter?  Isn't the get object routine simply going to one or the other but the data inside E+ is the same?  Otherwise, wouldn't every object have two different paths?  So, if this fix is simply looking at what has already been read in and then simply choosing to process the already read data into local arrays in a slightly different order, what does it matter whether the data came from a JSON file or an IDF?  I understand the concept of not wanting to have to come back and redo this later, but given what I am proposing to do here, I don't comprehend why this would need to be redone later.  Of course, I might not understand something about what JSON is doing.  Thanks for your feedback.

RKStrand committed:
Revised AFN Linkage Input Order Fix

The previous solution worked for the current version, but with the move to JSON for input, a reviewer expressed concerns that this would have to be fixed when this input was transitioned to JSON.  Since JSON reorders the input, a more comprehensive solution was needed.  This commit converts the AirflowNetwork:Distribution:Linkage input to using JSON directly and the solution should work for any input that includes these objects in any order.  So, conversion to JSON for this object is now complete as part of the solution.

RKStrand commented:
@Myoldmopar Thanks so much for that explanation!  I think I understand much better now and I can see where doing it for this fix would give me a chance to figure this out on a small scale so that in future work I'll be a little more up to speed.  I have marked this as "waiting for developer" and will look to implement what @jasondegraw requested.  (Oops...this was supposed to go out several days ago, but I forgot to click "Comment")

jasondegraw reviewed (approved):
@RKStrand This will be better long term, thanks for being willing to go back over it.

Myoldmopar committed:
Merge develop; resolve conflicts

rraustad commented on src/EnergyPlus/AirflowNetwork/src/Solver.cpp:
```diff
@@ -4838,96 +4838,98 @@ namespace AirflowNetwork {
+                    AirflowNetworkLinkageData(count).ZoneName =
+                        Util::makeUPPER(m_state.dataInputProcessing->inputProcessor->getAlphaFieldValue(fields, schemaProps, "thermal_zone_name"));
+                    std::string tempZone = AirflowNetworkLinkageData(count).ZoneName;
+                    AirflowNetworkLinkageData(count).LinkNum = count;
```
This conversion to json looks correct. There are some debugging statements sprinkled through (e.g., std::string tempName) that could be removed. The example files all pass and one unit test needs investigation/correction.

RKStrand replied:
@rraustad I got rid of the temporary assignments and made corrections that should result in the unit tests all passing now I think.  Also, the divide by zero noted by @Myoldmopar on 9/4 above, I fixed this by changing SupplyBranchArea in line 12808 in Solver.cc to ReturnTrunkArea.  From a quick glance at the code, I think this was just a typo that somehow never got tripped.  But given that ReturnTrunkArea is calculated right above in line 12807, it would seem likely that this fix is legitimate.  Thoughts?  If you are all good with this and it comes back green, then I think we are ready for the final review and hopefully accept/merge.

RKStrand committed:
Corrections for Unit Test Issues

Made some corrections to address issues that were present in the unit tests.  Note to @rraustad: not 100% sure about the fix to line 12808/12803 but it seems wrong in the code?  Also got rid of some temporary assignments that were added during debugging.

RKStrand committed:
Merge branch 'develop' into 10652OrderDependencyFixAFN

RKStrand commented:
@mitchute I just resolved the conflict I think.  I'm not sure why there was a conflict as I don't think I messed with anything here, but hopefully everything is fine now.

RKStrand committed:
Merge branch 'develop' into 10652OrderDependencyFixAFN

RKStrand committed:
Merge branch '10652OrderDependencyFixAFN' of https://github.com/NREL/EnergyPlus into 10652OrderDependencyFixAFN

RKStrand committed:
Correction for Input File Fail

Interesting.  Forgot to convert the name to uppercase which surprisingly became a problem only when dealing with duct loss stuff.

mitchute committed:
Merge branch 'develop' into 10652OrderDependencyFixAFN

RKStrand committed:
Get rid of the changes in unused object count

Hopefully this does the trick...

mitchute commented:
@RKStrand yeah, I think you can probably fix it up quick enough. I took the `makeUPPER` off of [this line](https://github.com/NREL/EnergyPlus/pull/11148/files#diff-e1efb8d61457e60529afd8b89b7ff9f3770dc9f1e1fbfba128dfaa82d5a3078eR4893), which cleaned up that warning, but now there are a couple other failures to figure out. I'll leave it to you for now.

RKStrand commented:
@mitchute Ok, I've pulled the latest with develop merged in and made that change.  Hoping for a Halloween treat here and no more "tricks"!

rraustad commented:
I was wondering why there were any diffs at all from these changes and then I reviewed and remembered you did this:

<img width="822" height="193" alt="{CC8A19C4-B51E-4983-8160-11E3F313892B}" src="https://github.com/user-attachments/assets/a0373c6d-db29-4615-ad5f-4d8580388c51" />
